rcl_logging_interface/src/logging_dir.c

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcutils/allocator.h>
+#include <rcutils/error_handling.h>
+#include <rcutils/filesystem.h>
+#include <rcutils/format_string.h>
+#include <rcutils/get_env.h>
+#include <rcutils/strdup.h>
+
+#include "rcl_logging_interface/rcl_logging_interface.h"
+
+rcl_logging_ret_t
+rcl_logging_get_logging_directory(rcutils_allocator_t allocator, char ** directory)
+{
+  if (NULL == directory) {
+    RCUTILS_SET_ERROR_MSG("directory argument must not be null");
+    return RCL_LOGGING_RET_INVALID_ARGUMENT;
+  }
+  if (NULL != *directory) {
+    RCUTILS_SET_ERROR_MSG("directory argument must point to null");
+    return RCL_LOGGING_RET_INVALID_ARGUMENT;
+  }
+
+  const char * log_dir_env;
+  const char * err = rcutils_get_env("ROS_LOG_DIR", &log_dir_env);
+  if (NULL != err) {
+    RCUTILS_SET_ERROR_MSG("rcutils_get_env failed");
+    return RCL_LOGGING_RET_ERROR;
+  }
+  if ('\0' != *log_dir_env) {
+    *directory = rcutils_strdup(log_dir_env, allocator);
+    if (NULL == *directory) {
+      RCUTILS_SET_ERROR_MSG("rcutils_strdup failed");
+      return RCL_LOGGING_RET_ERROR;
+    }
+  } else {
+    const char * ros_home_dir_env;
+    err = rcutils_get_env("ROS_HOME", &ros_home_dir_env);
+    if (NULL != err) {
+      RCUTILS_SET_ERROR_MSG("rcutils_get_env failed");
+      return RCL_LOGGING_RET_ERROR;
+    }
+    if ('\0' == *ros_home_dir_env) {
+      ros_home_dir_env = rcutils_join_path("~", ".ros", allocator);
+      if (NULL == ros_home_dir_env) {
+        RCUTILS_SET_ERROR_MSG("rcutils_join_path failed");
+        return RCL_LOGGING_RET_ERROR;
+      }
+    }
+    *directory = rcutils_join_path(ros_home_dir_env, "log", allocator);
+    if (NULL == *directory) {
+      RCUTILS_SET_ERROR_MSG("rcutils_join_path failed");
```

Comment:
There are four memory problems here:

1.  If this second `rcutils_join_path` fails, but `ros_home_dir_env` was not set in the environment, then we leak memory.
1.  `ros_home_dir_env` is never freed if it was allocated.
1.  In the case where we are expanding the `~`, we are leaking the original memory allocated to `directory`.
1.  `ros_home_dir_env` is captured two different ways; via a memory allocation (in the case where `ROS_HOME` is not set), and via a static pointer (in the case where `ROS_HOME` is set).  This makes it impossible to do the right thing to cleanup.

All of these need to be cleaned up.  I *might* suggest splitting this function into multiple functions; I think that might make it easier to determine what should happen in each case.

Reply:
Uff, I turned off my memory checker when reviewing this :joy:.

Reply:
Yeah, the way it's written really doesn't make it easy. I'll go over it once again.

> 3. In the case where we are expanding the ~, we are leaking the original memory allocated to directory.

It's freed on line 83 with `directory_not_expanded` (unless I'm missing something).

Reply:
Also maybe we should include https://github.com/ros2/rcutils/pull/298 in this PR after all :sweat_smile:

Reply:
> It's freed on line 83 with `directory_not_expanded` (unless I'm missing something).

Oh, you are totally right, sorry about that.

Reply:
> Yeah, the way it's written really doesn't make it easy. I'll go over it once again. Sorry for all the trouble.

No worries, this kind of code can be hard to get right in C.  In particular, we have to go through and be careful about the memory lifetimes here, since some of them are static and some of them are heap.  I might suggest doing something (an extra copy or something like that) to make sure they are all heap before the expansion steps.  That way it is easy to just always free it at the end.

Reply:
fixed in 0dd91f4655e7b036337ab62826987aca55873984

I went with the "extra copy" approach (and hopefully made it more obvious); I think that should resolve all the problems you mentioned above @clalancette

Reply:
This part is looking better, thanks.